### ISSUE TITLE
MMCIF2Dict ValueError: Opening quote in middle of word

### ISSUE THREAD

**Issue by andrei-istrate** [biopython/biopython#2261]
### Setup
I am reporting a problem with Biopython version, Python version, and operating
system as follows:

import sys; print(sys.version)
3.6.4 (v3.6.4:d48ecebad5, Dec 18 2017, 21:07:28)
[GCC 4.2.1 (Apple Inc. build 5666) (dot 3)]
import platform; print(platform.python_implementation()); print(platform.platform())
CPython
Darwin-18.5.0-x86_64-i386-64bit
import Bio; print(Bio.version)
1.74

Please correct me if I am wrong but I think that the _splitline() function in MMCIF2Dict module does not behave completely in agreement with the syntax defined on https://www.iucr.org/resources/cif/spec/version1.1/cifsyntax.

Specifically point 14 says “Matching single or double quote characters (' or ") may be used to bound a string representing a non-simple data value provided the string does not extend over more than one line.” It does not say that ' or " can’t be used in a simple data value. 

This means that:
“foo bar’ baz” 
is a valid cif line and the following example:

from Bio.PDB.MMCIF2Dict import MMCIF2Dict
d = MMCIF2Dict('any_valid_cif_file')
print(list(d._splitline(“foo bar’ baz”)))

and should give:
[‘foo’, “bar’”, ‘baz’]

### Actual behaviour
Traceback (most recent call last):
  File "/Users/andrei/PycharmProjects/3D-STRUDEL/playground/tmp.py", line 3, in <module>
    print(list(d._splitline_orig("foo bar' baz")))
  File "/Users/andrei/PycharmProjects/3D-STRUDEL/venv/lib/python3.6/site-packages/Bio/PDB/MMCIF2Dict.py", line 87, in _splitline_orig
    raise ValueError("Opening quote in middle of word: " + line)
ValueError: Opening quote in middle of word: foo bar' baz

This is a problem because at least one widely used program (REFMAC) writes atom_site lines like this:
ATOM    55932 C  C5'  . A   2   46    21   C5'  . A     21  120.670   118.398   132.018 1.000   124.517 .    21    A   2 C5'  .   1 

Which is not compatible with the current version of MMCIF2Dict

I have slightly modified the _splitline() method to accommodate this

    quote_chars = ['\'', '\"']
    whitespace_chars = [' ', '\t']    
    
    def _splitline(line):
        # See https://www.iucr.org/resources/cif/spec/version1.1/cifsyntax for the syntax
        in_token = False
        # quote character of the currently open quote, or None if no quote open
        quote_open_char = None
        start_i = 0
        for (i, c) in enumerate(line):
            if c in whitespace_chars:
                if in_token and not quote_open_char:
                    in_token = False
                    yield line[start_i:i]
            elif c in quote_chars:
                if not quote_open_char and not in_token:
                    quote_open_char = c
                    in_token = True
                    start_i = i + 1
                elif c == quote_open_char and (i + 1 == len(line) or line[i + 1] in whitespace_chars):
                    quote_open_char = None
                    in_token = False
                    yield line[start_i:i]
            elif c == "#" and not in_token:
                # Skip comments. "#" is a valid non-comment char inside of a
                # quote and inside of an unquoted token (!?!?), so we need to
                # check that the current char is not in a token.
                return
            elif not in_token:
                in_token = True
                start_i = i
        if in_token:
            yield line[start_i:]
        if quote_open_char:
            raise ValueError("Line ended with quote open: " + line)

    print(list(_splitline("foo bar' baz")))

the result is:
[‘foo’, “bar’”, ‘baz’]
which adds compatibility with REFMAC and probably some other programs.

**Comment by peterjc** [comment]
Turning your suggestion into a pull request would be helpful - we'd get to see if this has any side effecting in the test suite etc.

**Comment by peterjc** [comment]
@jgreener64 could you look at this please?

**Comment by jgreener64** [comment]
Thanks for raising this. Initially I thought you were correct, though on digging into it I'm not so sure.

The spec also says:
>15. Because data values are invariably separated from other tokens in the file by white space, such a quote-delimited character string may contain instances of the character used to delimit the string provided they are not followed by white space. For example, the data item
>
>     _example  'a dog's life' 
>
>is legal; the data value is a dog's life.

It is ambiguous whether "followed by white space" means "followed by white space in the data value" (in which case ending with a quote is okay) or "followed by a white space in the file" (in which case ending with a quote is not okay). I would prefer to err on the side of caution and only allow quoted versions of these cases, i.e. `"C5'"`.

The  PDB puts quotes around the types of examples you give, for example:
```
ATOM   1   O "O5'" . DC  A 1 1 ? 84.780 76.459 3.055   1.00 48.65 ? 1   DC  A "O5'" 1 
```
is line 759 of the mmCIF file for 6GN2.

Either they are doing this because it is required by the format, or they are being conservative. Either way I think we should agree with them and not allow your example to parse. I am willing to be convinced otherwise though.

**Comment by peterjc** [comment]
Can you report this to REFMAC, highlighting that the PDB wrap such values with double quotes, and see how they view things?

**Comment by andrei-istrate** [comment]
I talked with Garib Murshudov (the main developer of REFMAC) and he thinks that wrapping such cases in double quotes is redundant and reduces flexibility so he would prefer to keep it as it is.

**Comment by peterjc** [comment]
Do we have any contacts that the PDB who could clarify the spec's intended meaning?

**Comment by andrei-istrate** [comment]
I can talk with PDB people.

**Comment by epeisach** [comment]
If you are doing development work - here is an adjustment to the test suite to test this case... 

Speaking from the PDB point of view - we know there are parsers that have issues with edge cases - and in order to be conservative - we are quoting such instances. There is nothing wrong with extra quoting. Other edge cases exist in other parsers - such as  ms#29. Some parsers would treat the # as the start of a comment. (Biopython fixed this April 2018). However, CIF started within the IUCR context. 

Personally, I always go back to the formal grammar in the appendix and I do not see an ambiguity of an <UnquotedString> that contains O5' as this needs to be an <OrdinaryChar> while the rest are <NonblankChar> - NonBlankChar includes a single quote.  The discussion of a whitespace - a whitespace includes an <eol>, tab or space character.  So unless the file ends without a newline, tab or space, you will have such an item before another value, comment, etc. So strictly speaking - if the file ends O5' -- with no newline - you probably need to quote. 

 

diff --git a/Tests/test_PDB_MMCIF2Dict.py b/Tests/test_PDB_MMCIF2Dict.py
index b34af6043..b9d8203b6 100644
--- a/Tests/test_PDB_MMCIF2Dict.py
+++ b/Tests/test_PDB_MMCIF2Dict.py
@@ -69,6 +69,8 @@ class MMCIF2dictTests(unittest.TestCase):
         self.assertEqual(list(mmcif._splitline("foo 'bar'a' b")), ["foo", "bar'a", "b"])
         self.assertEqual(list(mmcif._splitline("foo \"bar' a\" b")), ["foo", "bar' a", "b"])
         self.assertEqual(list(mmcif._splitline("foo '' b")), ["foo", "", "b"])
+        self.assertEqual(list(mmcif._splitline('foo "O5\'" b')), ["foo", "O5'", "b"])
+        self.assertEqual(list(mmcif._splitline('foo O5\' b')), ["foo", "O5'", "b"])
 
         # A hash (#) starts a comment iff it is preceded by whitespace or is at
         # the beginning of a line:

**Comment by peterjc** [comment]
Fixed via #2716, closing - thank you!
